### PR TITLE
Improved caching for better load times

### PR DESCRIPTION
**Pull request by LXIF**
Added file hashes to chunk and asset filenames, as well as a 'cacheable' tag since we are setting our headers in the certify_asset function and the cacheable tag lets us easily distinguish cacheable files.

Since the relevant files now include their filehash in their name, they can be cached long-term, as any file change will change their name and cause them to be reloaded.

<hr/><details><summary>🟡 Some screens were changed</summary></details>

### CODE REVIEW

**Comment by LXIF** [comment]
@lmuntaner The tests all pass but the failing ones are because of the following:

- filenames now include file hashes, so the paths in the integration tests that access those files can no longer be hardcoded
- i am looking for their names in the dist directory
- dist directory seems to not exist in our integration test environment on github

do you have any idea how to fix this?

**Comment by LXIF** [comment]
@lmuntaner fixed it, but for some reason now the response certification for the spa file doesn't go through unless i lower the certification version to 1. Do you have expertise in this field^^

**Comment by lmuntaner** [comment]
> @lmuntaner fixed it, but for some reason now the response certification for the spa file doesn't go through unless i lower the certification version to 1. Do you have expertise in this field^^

I don't understand why the spa.js passes only with v1 certification. But if I'm not mistaken, the v1 doesn't certify headers. Therefore, they won't be passed to the browser. Better ask @nathanosdev , I believe he's the one that will explain it better.

**Review by lmuntaner — COMMENTED** [review]
Thanks!! Did you try deploying this to mainnet and see that it works as expected?

You can read the different ways to deploy to mainnet for testing [here](https://www.notion.so/dfinityorg/Deploy-II-to-Mainnet-for-Testing-90d63632cb4b4146818c442ea5761457?pvs=4).

Just a few comments.

**Comment by nathanosdev** [comment]
I haven't dug into the code as to why response verification v2 is failing, because I have an alternative suggestion. Unless this PR is really time sensitive, could you guys hold off for a couple of weeks until we release the `ic-asset-certification` library? Then you can actually just remove all of this custom implementation and will likely have a much easier time implementing these changes.

**Comment by LXIF** [comment]
> I haven't dug into the code as to why response verification v2 is failing, because I have an alternative suggestion. Unless this PR is really time sensitive, could you guys hold off for a couple of weeks until we release the `ic-asset-certification` library? Then you can actually just remove all of this custom implementation and will likely have a much easier time implementing these changes.

@nathanosdev Thanks for checking in! I'm assuming this library takes over asset certification and verification? This PR is primarily about browser caching and doesn't change anything about how we do certification and verification per se, the question just arose in the integration test (and strangely it only arises for the spa.js file, not for any of the other files like fonts or css). IMO the UX improvements from browser caching are worth deploying now, but we'd of course be happy to implement ```ic-asset-certification``` when it's out in a separate PR. Am open to opposing viewpoints though.

File: src/internet_identity/tests/integration/http.rs [hunk 1]
```diff
@@ -123,7 +182,69 @@ fn should_set_cache_control_for_fonts() -> Result<(), CallError> {
...
+    };
+    let index_response = http_request(&env, canister_id, &index_request)?;
+
+    // Convert body to string and find the font URL
+    let html = String::from_utf8(index_response.body.into_vec()).expect("Failed to parse HTML");
+
+    // Find the spa URL in the HTML
+    let spa_url = {
+        let spa_suffix = "cacheable.js";
+        let spa_end = html
```
[the remark this answers is not in the document]
Comment (LXIF) [reply src/internet_identity/tests/integration/http.rs:223]: Currently yes, but it's the only one there. Can refine.

**Comment by nathanosdev** [comment]
> I'm assuming this library takes over asset certification and verification? 

Yep! If you're interested, you can see how it looks to use the library in [this example project](https://github.com/dfinity/response-verification/tree/main/examples/http-certification/assets).

> This PR is primarily about browser caching and doesn't change anything about how we do certification and verification per se

I understand that's the _intention_ of the PR, but something has broken v2 certification, which is required for the caching to work. Without a proper v2 certification, the cache headers [are stripped out](https://github.com/dfinity/http-gateway/blob/main/packages/ic-http-gateway/src/protocol/handler.rs#L283-L288) to prevent malicious replicas from dangerously inserting cache headers where they don't belong (on the `index.html` file for example).

So my suggestion was made in the context of this comment:
> for some reason now the response certification for the spa file doesn't go through unless i lower the certification version to 1

So rather than dig through the code to figure out why that's broken, I suggested to just use the library once it's released. Of course if you think it's worth while to find the source of the issue in order to release this UX improvement sooner then that makes sense to me.

File: src/internet_identity/tests/integration/http.rs [hunk 2]
```diff
@@ -111,9 +111,68 @@ fn should_set_cache_control_for_fonts() -> Result<(), CallError> {
...
+            .expect("Could not find cacheable.css in HTML");
+        let prefix_start = html[..css_end]
+            .rfind('/')
+            .expect("Could not find starting / for CSS URL");
+
+        html[prefix_start..css_end + css_suffix.len()].to_string()
+    };
+
+    //get css file
+    let css_request = HttpRequest {
```
[the remark this answers is not in the document]
Comment (LXIF) [reply src/internet_identity/tests/integration/http.rs:141]: Our index html only has three link tags - one for the favicon, one for the spa.css and one for the spa.js, so for now we're safe by just going by tag. I can add additional checks to make sure it's the correct one, though.

**Comment by LXIF** [comment]
Alright, will see if I can find something within a reasonable timeframe, otherwise will wait :)

File: src/internet_identity/tests/integration/http.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lmuntaner) [reply src/internet_identity/tests/integration/http.rs:223]: Let's add a comment about assuming only three. I can see use changing this and suddenly the test will fail. Or worse, it won't fail but not test what we want it to test.

**Comment by LXIF** [comment]
> I understand that's the _intention_ of the PR, but something has broken v2 certification, which is required for the caching to work. Without a proper v2 certification, the cache headers [are stripped out](https://github.com/dfinity/http-gateway/blob/main/packages/ic-http-gateway/src/protocol/handler.rs#L283-L288) to prevent malicious replicas from dangerously inserting cache headers where they don't belong (on the `index.html` file for example).

@nathanosdev Quick additional question - the headers do seem to work when testing in locally deployed dfx (the files in question do get stored in cache). Would it be correct to assume that this means that v2 certification does work but my tests are broken somehow?

**Comment by LXIF** [comment]
@nathanosdev never mind, found the (very stupid) mistake 😂

**Comment by nathanosdev** [comment]
> never mind, found the (very stupid) mistake 😂

Awesome 👏 I'll let you know later anyway once the library is ready!

**Review by lmuntaner — APPROVED** [review]
Thanks!

**Pull request merged**